rdma-ndd/rdma-ndd.c

```diff
@@ -0,0 +1,462 @@
+			device);
+	nd_file[sizeof(nd_file)-1] = '\0';
+
+	f = fopen(nd_file, "r+");
+	if (!f) {
+		syslog(LOG_ERR, "Failed to open %s\n", nd_file);
+		return -EIO;
+	}
+
+	if (!fgets(nd, sizeof(nd), f)) {
+		syslog(LOG_ERR, "Failed to read %s\n", nd_file);
+		rc = -EIO;
+		goto error;
+	}
+	newline_to_null(nd);
+
+	build_node_desc(new_nd, sizeof(new_nd), device, hostname);
+
+	if (!force && strncmp(new_nd, nd, sizeof(new_nd)) == 0) {
+		syslog(LOG_INFO, "%s: no change (%s)\n", device, new_nd);
+	} else {
+		syslog(LOG_INFO, "%s: change (%s) -> (%s)\n",
+			device, nd, new_nd);
+		rewind(f);
+		fprintf(f, "%s", new_nd);
+	}
+
+	rc = 0;
+error:
+	fclose(f);
+	return rc;
+}
+
+static int set_rdma_node_desc(const char *hostname, int force)
+{
+	DIR *class_dir;
+	struct dirent *dent;
+
+	class_dir = opendir(SYS_INFINIBAND);
+	if (!class_dir) {
+		syslog(LOG_INFO, "Failed to open " SYS_INFINIBAND);
+		return -ENOSYS;
+	}
+
+	while ((dent = readdir(class_dir))) {
+		int retry = set_retry_cnt;
+		if (dent->d_name[0] == '.')
+			continue;
+
+		while (update_node_desc(dent->d_name, hostname, force) && retry > 0) {
+			syslog(LOG_ERR, "retrying set Node Description on %s\n",
+				dent->d_name);
+			retry--;
+		}
+	}
+
+	closedir(class_dir);
+	return 0;
+}
+
+static int read_hostname(int fd, char *name, size_t len)
+{
+	int rc;
+	memset(name, 0, len);
+	if (read(fd, name, len-1) >= 0) {
+		newline_to_null(name);
+		strip_domain(name);
+		rc = 0;
+	} else {
+		syslog(LOG_ERR, "Read %s Failed\n", SYS_HOSTNAME);
+		rc = -EIO;
+	}
+	return rc;
+}
+
+static void setup_udev(void)
+{
+	udev = udev_new();
+	if (!udev) {
+		syslog(LOG_ERR, "udev_new failed\n");
+		return;
+	}
+}
+
+static int get_udev_fd(void)
+{
+	mon = udev_monitor_new_from_netlink(udev, "udev");
+	if (!mon) {
+		syslog(LOG_ERR, "udev monitoring failed\n");
+		return -1;
+	}
+
+	udev_monitor_filter_add_match_subsystem_devtype(mon, "infiniband", NULL);
+	udev_monitor_enable_receiving(mon);
+	return udev_monitor_get_fd(mon);
+}
+
+static void process_udev_event(int ud_fd, const char *hostname)
+{
+	struct udev_device *dev;
+
+	dev = udev_monitor_receive_device(mon);
+	if (dev) {
+		const char *device = udev_device_get_sysname(dev);
+		const char *action = udev_device_get_action(dev);
+
+		syslog(LOG_INFO, "Device event: %s, %s, %s\n",
+			udev_device_get_subsystem(dev),
+			device, action);
+
+		if (device && action
+		    && strncmp(action, "add", sizeof("add")) == 0)
+			update_node_desc(device, hostname, 1);
+
+		udev_device_unref(dev);
+	}
+}
+
+static void monitor(void)
+{
+	char hostname[128];
+	int hn_fd;
+	int rc;
+	struct pollfd fds[2];
+	int numfds = 1;
+	int ud_fd;
+
+	ud_fd = get_udev_fd();
+	if (ud_fd >= 0)
+		numfds = 2;
+
+	while (1) {
+		hn_fd = open(SYS_HOSTNAME, O_RDONLY);
+		if (hn_fd < 0) {
+			syslog(LOG_ERR,
+				"Open %s Failed: retry in %d seconds\n",
+				SYS_HOSTNAME, failure_retry_rate);
+			sleep(failure_retry_rate);
```

Comment:
Don't include user tunable knobs that have no purpose. If the failure cannot happen then wire a reasonable defensive delay and be done with it. It is especially problematic to have some configurable in the config file and some on the command line. Be consistent, and eliminate things that do not actually need to be configured..